### PR TITLE
constexpr int/int constexpr to enum class continued

### PR DESCRIPTION
**Pull request by dareumnam**
Pull request overview
---------------------
 - `constexpr int` to `enum class` in
   - SimAirServingZones.hh
   - MixedAir.hh
   - DataWindowEquivalentLayer.hh 
   - DataHeatBalance.hh (partial)

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [X] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [X] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [X] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: src/EnergyPlus/DataHeatBalance.hh [hunk 1]
```diff
@@ -155,39 +142,68 @@ namespace DataHeatBalance {
...
+        SurfaceWeighted,
+        AngleFactor
+    };
 
     // Parameters for Ventilation
-    constexpr int NaturalVentilation(0);
-    constexpr int IntakeVentilation(1);
-    constexpr int ExhaustVentilation(2);
-    constexpr int BalancedVentilation(3);
+    enum class Vent
```
Comment (amirroth) [inline src/EnergyPlus/DataHeatBalance.hh]: Can probably call this `VentilationType` and remove the word `Ventilation` from the enumerated values.

File: src/EnergyPlus/SurfaceGeometry.cc [hunk 2]
```diff
@@ -7259,15 +7264,21 @@ namespace SurfaceGeometry {
 
             Roughness = state.dataIPShortCut->cAlphaArgs(3);
             // Select the correct Number for the associated ascii name for the roughness type
-            if (UtilityRoutines::SameString(Roughness, "VeryRough")) state.dataSurface->ExtVentedCavity(Item).BaffleRoughness = VeryRough;
-            if (UtilityRoutines::SameString(Roughness, "Rough")) state.dataSurface->ExtVentedCavity(Item).BaffleRoughness = Rough;
-            if (UtilityRoutines::SameString(Roughness, "MediumRough")) state.dataSurface->ExtVentedCavity(Item).BaffleRoughness = MediumRough;
-            if (UtilityRoutines::SameString(Roughness, "MediumSmooth")) state.dataSurface->ExtVentedCavity(Item).BaffleRoughness = MediumSmooth;
-            if (UtilityRoutines::SameString(Roughness, "Smooth")) state.dataSurface->ExtVentedCavity(Item).BaffleRoughness = Smooth;
-            if (UtilityRoutines::SameString(Roughness, "VerySmooth")) state.dataSurface->ExtVentedCavity(Item).BaffleRoughness = VerySmooth;
+            if (UtilityRoutines::SameString(Roughness, "VeryRough"))
```
Comment (amirroth) [inline src/EnergyPlus/SurfaceGeometry.cc]: This if-ladder (which should really be an if-else-if ladder, but that's besides the point) should really be a call to a function that loops over an array of `SurfaceRoughness` names and returns the index to it, which you can then static_cast to a `SurfaceRoughness`. 

Something like: 

`constexpr std::array<std::string_view, SurfaceRoughness::Num> SurfaceRoughnessNames = {"Very Rough", "Rough", ... }; // in the .hh file
`
`BaffleRoughness = static_cast<SurfaceRoughness>FindEnumeratedValueIndex(Roughness, SurfaceRoughnessNames);`

The presence of `gsl::span` should allows us to create a `FindEnumeratedValueIndex` that takes different `std::array`'s as arguments.

You don't have to do this in this PR, but put a TODO pin here for later.

File: src/EnergyPlus/SurfaceGeometry.cc [hunk 3]
```diff
@@ -11420,12 +11443,16 @@ namespace SurfaceGeometry {
                                                              "WindowMaterial:Screen:EquivalentLayer",
                                                              "WindowMaterial:Gap:EquivalentLayer"});
 
-                    int const MaterialLayerGroup = state.dataMaterial->Material(MaterNum).Group;
-                    if ((MaterialLayerGroup == WindowSimpleGlazing) || (MaterialLayerGroup == ShadeEquivalentLayer) ||
-                        (MaterialLayerGroup == DrapeEquivalentLayer) || (MaterialLayerGroup == BlindEquivalentLayer) ||
-                        (MaterialLayerGroup == ScreenEquivalentLayer) || (MaterialLayerGroup == GapEquivalentLayer)) {
+                    DataHeatBalance::MaterialGroup const MaterialLayerGroup = state.dataMaterial->Material(MaterNum).Group;
+                    if ((MaterialLayerGroup == DataHeatBalance::MaterialGroup::WindowSimpleGlazing) ||
```
Comment (amirroth) [inline src/EnergyPlus/SurfaceGeometry.cc]: Good spot for `BITF_TEST_ANY`!

File: src/EnergyPlus/EcoRoofManager.cc [hunk 4]
```diff
@@ -617,15 +617,15 @@ namespace EcoRoofManager {
                 Gammah = std::pow(1.0 - 5.0 * Rib, -0.5);
             }
 
-            if (RoughSurf == VerySmooth) { //  6= very smooth, 5=smooth, 4= med. sm. ,3= med. rough. , 2= rough, 1= Very rough
+            if (RoughSurf == DataSurfaces::SurfaceRoughness::VerySmooth) {
                 state.dataEcoRoofMgr->Zog = 0.0008;
```
Comment (amirroth) [inline src/EnergyPlus/EcoRoofManager.cc:621]: This should also be implemented as a lookup in a `std::array`.

File: src/EnergyPlus/DataAirSystems.hh [hunk 5]
```diff
@@ -90,18 +91,18 @@ namespace DataAirSystems {
...
-        int CompIndex;               // Component Index in whatever is using this component
-        HVACSystemData *compPointer; // pointer to HVAC system
-        int FlowCtrl;                // Component flow control (ACTIVE/PASSIVE)
-        bool ON;                     // When true, the designated component or operation scheme is available
-        bool Parent;                 // When true, the designated component is made up of sub-components
-        std::string NodeNameIn;      // Component inlet node name
-        std::string NodeNameOut;     // Component outlet node name
-        int NodeNumIn;               // Component inlet node number
-        int NodeNumOut;              // Component outlet node number
+        std::string TypeOf;                        // The 'keyWord' identifying  component type
```
Comment (amirroth) [inline src/EnergyPlus/DataAirSystems.hh:94]: Probably not part of this PR, but `std::string TypeOf` is redundant with `SimAirServingZones::CompType CompType_Num`. Only the second of these two should be needed.

File: src/EnergyPlus/DataHeatBalance.cc [hunk 6]
```diff
@@ -476,13 +483,20 @@ void CheckAndSetConstructionProperties(EnergyPlusData &state,
...
 
-            if (state.dataMaterial->Material(state.dataConstruction->Construct(ConstrNum).LayerPoint(1)).Group == Shade ||
-                state.dataMaterial->Material(state.dataConstruction->Construct(ConstrNum).LayerPoint(1)).Group == WindowGas ||
-                state.dataMaterial->Material(state.dataConstruction->Construct(ConstrNum).LayerPoint(1)).Group == WindowGasMixture ||
-                state.dataMaterial->Material(state.dataConstruction->Construct(ConstrNum).LayerPoint(1)).Group == WindowBlind ||
-                state.dataMaterial->Material(state.dataConstruction->Construct(ConstrNum).LayerPoint(1)).Group == Screen ||
-                state.dataMaterial->Material(state.dataConstruction->Construct(ConstrNum).LayerPoint(1)).Group == ComplexWindowShade ||
-                state.dataMaterial->Material(state.dataConstruction->Construct(ConstrNum).LayerPoint(1)).Group == ComplexWindowGap) {
+            if (state.dataMaterial->Material(state.dataConstruction->Construct(ConstrNum).LayerPoint(1)).Group ==
+                    DataHeatBalance::MaterialGroup::Shade ||
```
Comment (amirroth) [inline src/EnergyPlus/DataHeatBalance.cc]: Use `BITF_TEST_ANY`.

File: src/EnergyPlus/DataHeatBalance.cc [hunk 7]
```diff
@@ -499,13 +513,16 @@ void CheckAndSetConstructionProperties(EnergyPlusData &state,
...
             MaterNum = state.dataConstruction->Construct(ConstrNum).LayerPoint(Layer);
             if (MaterNum == 0) continue; // error -- has been caught will stop program later
-            if (state.dataMaterial->Material(MaterNum).Group == WindowGlass) ++TotGlassLayers;
-            if (state.dataMaterial->Material(MaterNum).Group == WindowSimpleGlazing) ++TotGlassLayers;
-            if (state.dataMaterial->Material(MaterNum).Group == Shade || state.dataMaterial->Material(MaterNum).Group == WindowBlind ||
-                state.dataMaterial->Material(MaterNum).Group == Screen || state.dataMaterial->Material(MaterNum).Group == ComplexWindowShade)
+            if (state.dataMaterial->Material(MaterNum).Group == DataHeatBalance::MaterialGroup::WindowGlass) ++TotGlassLayers;
+            if (state.dataMaterial->Material(MaterNum).Group == DataHeatBalance::MaterialGroup::WindowSimpleGlazing) ++TotGlassLayers;
+            if (state.dataMaterial->Material(MaterNum).Group == DataHeatBalance::MaterialGroup::Shade ||
+                state.dataMaterial->Material(MaterNum).Group == DataHeatBalance::MaterialGroup::WindowBlind ||
```
Comment (amirroth) [inline src/EnergyPlus/DataHeatBalance.cc]: `BITF_TEST_ANY`.

File: src/EnergyPlus/DataHeatBalance.cc [hunk 8]
```diff
@@ -575,19 +596,27 @@ void CheckAndSetConstructionProperties(EnergyPlusData &state,
...
         // Consistency checks for a construction with a between-glass shade or blind
 
-        if (TotShadeLayers == 1 && state.dataMaterial->Material(state.dataConstruction->Construct(ConstrNum).LayerPoint(1)).Group != Shade &&
-            state.dataMaterial->Material(state.dataConstruction->Construct(ConstrNum).LayerPoint(1)).Group != WindowBlind &&
-            state.dataMaterial->Material(state.dataConstruction->Construct(ConstrNum).LayerPoint(1)).Group != Screen &&
-            state.dataMaterial->Material(state.dataConstruction->Construct(ConstrNum).LayerPoint(TotLayers)).Group != Shade &&
-            state.dataMaterial->Material(state.dataConstruction->Construct(ConstrNum).LayerPoint(TotLayers)).Group != WindowBlind &&
-            state.dataMaterial->Material(state.dataConstruction->Construct(ConstrNum).LayerPoint(TotLayers)).Group != ComplexWindowShade &&
+        if (TotShadeLayers == 1 &&
+            state.dataMaterial->Material(state.dataConstruction->Construct(ConstrNum).LayerPoint(1)).Group != DataHeatBalance::MaterialGroup::Shade &&
```
Comment (amirroth) [inline src/EnergyPlus/DataHeatBalance.cc:599]: `BITF_TEST_ANY`.

File: src/EnergyPlus/SurfaceGeometry.cc (hunk 2 above)
Comment (dareumnam) [reply src/EnergyPlus/SurfaceGeometry.cc]: I'll do this on the next commit. Thanks!

File: src/EnergyPlus/DataAirSystems.hh (hunk 5 above)
Comment (dareumnam) [reply src/EnergyPlus/DataAirSystems.hh:94]: I got it.

File: src/EnergyPlus/DataHeatBalance.hh [hunk 9]
```diff
@@ -155,39 +142,68 @@ namespace DataHeatBalance {
...
-    constexpr int ZoneAveraged(1);
-    constexpr int SurfaceWeighted(2);
-    constexpr int AngleFactor(3);
+    enum class CalcMRT
+    {
+        Unassigned = -1,
+        ZoneAveraged,
+        SurfaceWeighted,
+        AngleFactor
+    };
```
Comment (mitchute) [inline src/EnergyPlus/DataHeatBalance.hh:168]: You're doing a ```static_cast<int>(mrtCalcType)``` in SQLiteProcedures.cc which uses these enum values. It looks like you aren't actually required to populate all of enum values for this to work as expected (see [Compiler Explorer example](https://godbolt.org/z/ojY7r9566)). However, do we want to explicitly fill them in anyways? 

@Myoldmopar @amirroth

**Review by mitchute — COMMENTED** [review]
This all looks OK to me. CI would complain if there were any issues.

@Myoldmopar are you concerned about the CI timeouts on the last commit?

File: src/EnergyPlus/ConvectionCoefficients.cc [hunk 10]
```diff
@@ -2473,11 +2474,12 @@ Real64 CalcASHRAESimpExtConvectCoeff(int const Roughness,       // Integer index
     Real64 CalcASHRAESimpExtConvectCoeff;
 
     // FUNCTION PARAMETER DEFINITIONS:
-    static Array1D<Real64> const D(6, {11.58, 12.49, 10.79, 8.23, 10.22, 8.23});
-    static Array1D<Real64> const E(6, {5.894, 4.065, 4.192, 4.00, 3.100, 3.33});
-    static Array1D<Real64> const F(6, {0.0, 0.028, 0.0, -0.057, 0.0, -0.036});
+    static Array1D<Real64> const D({0, 5}, {11.58, 12.49, 10.79, 8.23, 10.22, 8.23});
+    static Array1D<Real64> const E({0, 5}, {5.894, 4.065, 4.192, 4.00, 3.100, 3.33});
+    static Array1D<Real64> const F({0, 5}, {0.0, 0.028, 0.0, -0.057, 0.0, -0.036});
```
Comment (Myoldmopar) [inline src/EnergyPlus/ConvectionCoefficients.cc:2479]: I guess maybe there was a reason these couldn't just be `std::array`?  I'll probably find the reason as I read along.

**Review by Myoldmopar — APPROVED** [review]
Lots of good cleanups in here.  The one thing I spotted was a potential spot to convert some fixed-size Array1D's into std::arrays.  I didn't see why they weren't converted, and I'm not sure it's enough to hold up this PR.

File: src/EnergyPlus/DataHeatBalance.hh (hunk 9 above)
Comment (Myoldmopar) [reply src/EnergyPlus/DataHeatBalance.hh:168]: I think that comes down to style.  If you explicitly list out the values, then you definitely know what they are, but then on longer lists, you are now adding something to maintain, rather than just letting the compiler assign them values internally.  I'm not getting a strong instinct one way or the other.

**Comment by Myoldmopar** [comment]
> @Myoldmopar are you concerned about the CI timeouts on the last commit?

The timeouts weren't happening on the penultimate commit, and nothing meaningful changed on the last one, so I think it was just the CI machine probably getting bogged down or something, I'm not worried.

**Comment by Myoldmopar** [comment]
CI is otherwise super happy here.  I'll hold the door open for a few minutes to let people get a final look if they want, and then I'll regret doing that when the comments start coming in, :shrug:   It's not like this is the final step in the enum work, so follow-up things can be performed in the next PR.

File: src/EnergyPlus/DataHeatBalance.hh (hunk 9 above)
Comment (mitchute) [reply src/EnergyPlus/DataHeatBalance.hh:168]: I'm inclined to leave it as-is and let the compiler figure it out. One potential issue with explicitly listing them out is that C++ is happy to let you have (by mistake or otherwise) the following:

```C++
enum class MyEnum
{
   Invalid = -1,
   A = 1,
   B = 1,
}
```

Then ```MyEnum::A == MyEnum::B``` evaluates to True. See another [Compiler Explorer example](https://godbolt.org/z/43dPzeonx). This seems to me like it's going to be prone to errors.

One more question: do we need to inherit from ```int``` on all of these? i.e. ```enum class MyEnum : int {}```. It seems the answer is "no", but I just want to double-check.

Comment (Myoldmopar) [reply src/EnergyPlus/DataHeatBalance.hh:168]: It defaults to int, so if we are going with int then we shouldn't need that listed anywhere.  I wonder if we should consider using `char`?    That's for a different day...

> Each enumeration also has an underlying type. The underlying type can be explicitly specified using an enum-base. For a scoped enumeration type, the underlying type is int if it is not explicitly specified. In both of these cases, the underlying 
 type is said to be fixed. 

[_C++ 20 Standard_](https://isocpp.org/files/papers/N4860.pdf)

**Comment by mitchute** [comment]
Thanks all for the comments. This is ready as-is. Merging.

File: src/EnergyPlus/ConvectionCoefficients.cc (hunk 10 above)
Comment (dareumnam) [reply src/EnergyPlus/ConvectionCoefficients.cc:2479]: I'll change these to std::arrays in the next PR. Thanks for spotting these!
